### PR TITLE
feat(npm): add support for lockfile v3

### LINKED ISSUE #476: Add support for NPM lockfileVersion v3
GitHub recently pushed updates to their runners, upgrading their npm version from  `8.19.3` to `9.3.1`[^1]. It seems also that Azure DevOps has updated their runners. With NPM 9, the `package-lock.json` format has changed by default, and it is not backwards compatible with other version of NPM. Running `npm install`, with NPM 9 automatically upgrades the `package-lock.json` to lockfile version 3.

> `"lockfileVersion": 3` The lockfile version used by npm v7, without backwards compatibility affordances. This is used for the hidden lockfile at node_modules/.package-lock.json, and will likely be used in a future version of npm, once support for npm v6 is no longer relevant. [^2]

When CD attempts to process a `package-lock.json` with a v3 lockfile, we throw an exception because of the new `package-lock.json` format:

```
[09:20:18 INF] Could not parse Jtokens from C:\Users\justinperez\Documents\RPE.Partner.Service.Customers.UI\package-lock.json file.
System.NullReferenceException: Object reference not set to an instance of an object.
   at Microsoft.ComponentDetection.Detectors.Npm.NpmComponentDetectorWithRoots.ProcessIndividualPackageJTokens(ISingleFileComponentRecorder singleFileComponentRecorder, JToken packageLockJToken, IEnumerable`1 packageJsonComponentStream, Boolean skipValidation) in C:\component-detection\src\Microsoft.ComponentDe
tection.Detectors\npm\NpmComponentDetectorWithRoots.cs:line 153
   at Microsoft.ComponentDetection.Detectors.Npm.NpmComponentDetectorWithRoots.<>c__DisplayClass28_0.<OnFileFoundAsync>b__1(JToken token) in C:\component-detection\src\Microsoft.ComponentDetection.Detectors\npm\NpmComponentDetectorWithRoots.cs:line 120
   at Microsoft.ComponentDetection.Detectors.Npm.NpmComponentDetectorWithRoots.ProcessAllPackageJTokensAsync(IComponentStream componentStream, JTokenProcessingDelegate jtokenProcessor) in C:\component-detection\src\Microsoft.ComponentDetection.Detectors\npm\NpmComponentDetectorWithRoots.cs:line 144
   at Microsoft.ComponentDetection.Detectors.Npm.NpmComponentDetectorWithRoots.SafeProcessAllPackageJTokensAsync(IComponentStream componentStream, JTokenProcessingDelegate jtokenProcessor) in C:\component-detection\src\Microsoft.ComponentDetection.Detectors\npm\NpmComponentDetectorWithRoots.cs:line 256

```

We need to update our NPM detector to handle the new v3 lockfile.

[^1]: https://github.com/actions/runner-images/commit/e1f0c3f22606d0a0535819d68ac6c7b05c752343
[^2]: https://docs.npmjs.com/cli/v7/configuring-npm/package-lock-json

### PR DESCRIPTION
**Pull request by melotic**
Closes #476

### CODE REVIEW

**Comment by melotic** [comment]
Open questions for the NPM team:

* We do not have access to the ancestors of a current node to determine if a component is nested into another one (_i.e._ `node_modules/x/node_modules/y`). So, we recursively find the shortest path to the root note to determine the ancestors. **Is there a possibility for multiple paths back to the root node?** Or if that's possible, will the component always be placed in the root? (_i.e._ `node_modules/y`)

File: src/Microsoft.ComponentDetection.Common/DependencyGraph/DependencyGraph.cs [hunk 1]
```diff
@@ -119,6 +119,25 @@ public IEnumerable<string> GetAllExplicitlyReferencedComponents()
             .Select(componentRefNode => componentRefNode.Id);
     }
 
+    public ICollection<string> GetAncestors(string componentId)
+    {
+        if (componentId == null)
+        {
+            throw new ArgumentNullException(nameof(componentId));
+        }
```
Comment (JamieMagee) [inline src/Microsoft.ComponentDetection.Common/DependencyGraph/DependencyGraph.cs:127]: nit:
```suggestion
        ArgumentNullException.ThrowIfNull(componentId);
```

File: src/Microsoft.ComponentDetection.Common/DependencyGraph/DependencyGraph.cs [hunk 2]
```diff
@@ -119,6 +119,25 @@ public IEnumerable<string> GetAllExplicitlyReferencedComponents()
...
+
+        if (!this.componentNodes.TryGetValue(componentId, out var componentRef))
+        {
+            // this component isnt in the graph, so it has no ancestors
```
Comment (JamieMagee) [inline src/Microsoft.ComponentDetection.Common/DependencyGraph/DependencyGraph.cs:131]:
```suggestion
            // this component isn't in the graph, so it has no ancestors
```

File: src/Microsoft.ComponentDetection.Common/DependencyGraph/DependencyGraph.cs [hunk 3]
```diff
@@ -119,6 +119,25 @@ public IEnumerable<string> GetAllExplicitlyReferencedComponents()
             .Select(componentRefNode => componentRefNode.Id);
     }
 
+    public ICollection<string> GetAncestors(string componentId)
```
Comment (JamieMagee) [inline src/Microsoft.ComponentDetection.Common/DependencyGraph/DependencyGraph.cs:122]: Can you add a test for this please?

File: src/Microsoft.ComponentDetection.Detectors/npm/NpmComponentDetectorWithRoots.cs [hunk 4]
```diff
@@ -316,31 +340,114 @@ private void EnqueueDependencies(Queue<(JProperty Dependency, TypedComponent Par
...
+            var inLock = false;
+            JProperty dependencyProperty;
+            foreach (var possibleDepPath in possibleDepPaths)
+            {
+                var ancestorNodeModulesPath = string.Format(
+                    "{0}/{1}/{0}/{2}",
+                    NodeModules,
+                    string.Join($"/{NodeModules}/", possibleDepPath),
+                    dependency.Name);
+
```
Comment (JamieMagee) [inline src/Microsoft.ComponentDetection.Detectors/npm/NpmComponentDetectorWithRoots.cs:423]: Is that second `NodeModules necessary? It looks like you would get `node_modules/node_modules/parent/node_modules/child`?

Comment (melotic) [reply src/Microsoft.ComponentDetection.Detectors/npm/NpmComponentDetectorWithRoots.cs:423]: Yes, its necessary. `string.Join` only places the string to be joined _between_ elements. So an array with 1 element doesn't get `node_modules` appended anywhere.

File: test/Microsoft.ComponentDetection.Detectors.Tests/NpmDetectorWithRootsTests.cs [hunk 5]
```diff
@@ -580,4 +606,71 @@ public async Task TestNpmDetector_DependencyGraphIsCreatedAsync()
         dependencyGraph.GetDependenciesForComponent(componentBId).Should().Contain(componentCId);
         dependencyGraph.GetDependenciesForComponent(componentCId).Should().HaveCount(0);
     }
+
+    [TestMethod]
+    public async Task TestNpmDetector_NestedNodeModulesV3Async()
```
Comment (JamieMagee) [inline test/Microsoft.ComponentDetection.Detectors.Tests/NpmDetectorWithRootsTests.cs:611]: Is this test case representative of a real `package-lock.json` file? Wouldn't component B get hoisted to the top level?

You also need another test case to cover the case
```
root -> package A -> package B (v2)
     \-> package B (v1)
```

Comment (melotic) [reply test/Microsoft.ComponentDetection.Detectors.Tests/NpmDetectorWithRootsTests.cs:611]: Yep, this was created by simply adding the `debug` package:

```json
{
  "name": "test-pkg",
  "version": "1.0.0",
  "lockfileVersion": 3,
  "requires": true,
  "packages": {
    "": {
      "name": "test-pkg",
      "version": "1.0.0",
      "license": "ISC",
      "dependencies": {
        "debug": "^4.3.4"
      }
    },
    "node_modules/debug": {
      "version": "4.3.4",
      "resolved": "https://registry.npmjs.org/debug/-/debug-4.3.4.tgz",
      "integrity": "sha512-PRWFHuSU3eDtQJPvnNY7Jcket1j0t5OuOsFzPPzsekD52Zl8qUfFIPEiswXqIvHWGVHOgX+7G/vCNNhehwxfkQ==",
      "dependencies": {
        "ms": "2.1.2"
      },
      // ...
    },
    "node_modules/ms": {
      "version": "2.1.2",
      "resolved": "https://registry.npmjs.org/ms/-/ms-2.1.2.tgz",
      "integrity": "sha512-sGkPx+VjMtmA6MX27oA4FBFELFCZZ4S4XqeGOXCv68tT+jb3vk/RyaKWP0PTKyWtmLSM0b+adUTEvbs1PEaH2w=="
    }
  }
}
```

Comment (melotic) [reply test/Microsoft.ComponentDetection.Detectors.Tests/NpmDetectorWithRootsTests.cs:611]: Yes, this configuration can be created by this sequence:

```bash
npm install send
npm install debug
npm remove send
```

This gives you this `package-lock.json`:

```json
{
  "name": "test-pkg",
  "version": "1.0.0",
  "lockfileVersion": 3,
  "requires": true,
  "packages": {
    "": {
      "name": "test-pkg",
      "version": "1.0.0",
      "license": "ISC",
      "dependencies": {
        "debug": "^4.3.4"
      }
    },
    "node_modules/debug": {
      "version": "4.3.4",
      "resolved": "https://registry.npmjs.org/debug/-/debug-4.3.4.tgz",
      "integrity": "sha512-PRWFHuSU3eDtQJPvnNY7Jcket1j0t5OuOsFzPPzsekD52Zl8qUfFIPEiswXqIvHWGVHOgX+7G/vCNNhehwxfkQ==",
      "dependencies": {
        "ms": "2.1.2"
      },
      // ...
    },
    "node_modules/debug/node_modules/ms": {
      "version": "2.1.2",
      "resolved": "https://registry.npmjs.org/ms/-/ms-2.1.2.tgz",
      "integrity": "sha512-sGkPx+VjMtmA6MX27oA4FBFELFCZZ4S4XqeGOXCv68tT+jb3vk/RyaKWP0PTKyWtmLSM0b+adUTEvbs1PEaH2w=="
    }
  }
}
```

**Review by JamieMagee — APPROVED** [review]

**Pull request merged**